cralloc/batch_alloc.go

```diff
@@ -0,0 +1,72 @@
+// Copyright 2024 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package cralloc
+
+import "sync"
+
+// BatchAllocator is used to allocate small objects in batches, reducing the
+// number of individual allocations.
+//
+// The tradeoff is that the lifetime of the objects in a batch are tied
+// together, which can potentially result in higher memory usage.
+//
+// Sample usage:
+//
+//	var someTypeBatchAlloc = MakeBatchAllocator[SomeType]()  // global
+//		...
+//		x := someTypeBatchAlloc.Alloc()
+type BatchAllocator[T any] struct {
+	pool sync.Pool
```

Comment:
Interesting to me to see a `sync.Pool` here. Could you add a comment about this choice? I think benefits are avoiding a mutex, and good locality for which socket the memory is accessed from? Anything else?
We don't control how many batches are going to be instantiated at any given point (usually on the order of GOMAXPROCS, but there's no strict limit) which could be a concern, though likely a theoretical one unless someone decides to us this with very large `T`.

Reply:
Adding a comment. Even though it's not its intended functionality, `sync.Pool` can be used to approximate per-cpu structures (some discussion in https://github.com/golang/go/issues/8281 and https://github.com/golang/go/issues/18802).

Reply:
Any unused objects from the pool will go away in two GC cycles, so we shouldn't hold on to them under memory pressure.